client/serviceregistration/workload.go

```diff
@@ -0,0 +1,92 @@
+package serviceregistration
+
+import (
+	"github.com/hashicorp/nomad/client/allocrunner/taskrunner/interfaces"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+// WorkloadServices describes services defined in either a Task or TaskGroup
+// that need to be syncronized with a service registration provider.
+type WorkloadServices struct {
+	AllocID string
+
+	// Name of the task and task group the services are defined for. For
+	// group based services, Task will be empty.
+	Task  string
+	Group string
```

Comment:
More nitpick: let's split these up, and instead emphasize Group is always set but Task is only set for task-level services

```suggestion
	// Group in which the service belongs for a group-level service, or the group in which task belongs for a task-level service.
	Group string
	
    // Task in which the service belongs for task-level service. Will be empty for a group-level service.
	Task  string
```